examples/audio/transformer_asr.py

```diff
@@ -0,0 +1,523 @@
+"""
+Title: Automatic Speech Recognition with Transformer
+Author: [Apoorv Nandan](https://twitter.com/NandanApoorv)
+Date created: 2021/01/13
+Last modified: 2021/01/13
+Description: Implement a complete transformer and train it on sequence to sequence ASR.
+"""
+"""
+## Introduction
+
+Automatic speech recognition (ASR) consists of transcribing audio segments into text.
+A popular way to handle this task is to treat it as a sequence-to-sequence
+problem. The audio can be represented as a sequence of feature vectors,
+and the text can be represented as a sequence of characters, words, or subword tokens.
+
+For this demonstration, we will use LJSpeech dataset from the
+[LibriVox](https://librivox.org/) project. It consists of short
+audio clips of a single speaker reading passages from 7 non-fiction books.
+Our model will be similar to the original Transformer (both encoder and decoder)
+as proposed in the paper, "Attention is All You Need".
+
+
+**References:**
+- [Attention is All You Need](https://papers.nips.cc/paper/2017/file/3f5ee243547dee91fbd053c1c4a845aa-Paper.pdf)
+- [Very Deep Self-Attention Networks for End-to-End Speech Recognition](https://arxiv.org/pdf/1904.13377.pdf)
+- [Speech Transformers](https://ieeexplore.ieee.org/document/8462506)
+- [LJSpeech Dataset](https://keithito.com/LJ-Speech-Dataset/)
```

Comment:
Suggestion: maybe this can help save some space:

```suggestion
For this demonstration, we will use [LJSpeech](https://keithito.com/LJ-Speech-Dataset/) dataset from the
[LibriVox](https://librivox.org/) project. It consists of short
audio clips of a single speaker reading passages from seven non-fiction books.
The model you will use is similar to the original encoder-decoder Transformer model,
as proposed in Attention is All You Need ([Vaswani et al., 2017](https://papers.nips.cc/paper/2017/file/3f5ee243547dee91fbd053c1c4a845aa-Paper.pdf)). You can also refer to Very Deep Self-Attention Networks for End-to-End Speech Recognition ([Pham et al, 2019](https://arxiv.org/pdf/1904.13377.pdf)) and Speech-Transformer: A No-Recurrence Sequence-to-Sequence Model for Speech Recognition ([Dong et al., 2018](https://ieeexplore.ieee.org/document/8462506)).
```

(Note the IEEE link is "paywalled".)